### PR TITLE
windows: improved handling of invalid FDs

### PR DESCRIPTION
**Pull request by orangemocha**
If passed and invalid FD, _get_osfhandle() sets an error code
through errno, not _doserrno. Hence we need to use
SET_REQ_WIN32_ERROR insted of SET_REQ_RESULT.

In debug builds, _get_osfhandle() also raises a superfluous
assert. I implemented a wrapper that disables all asserts
around the call to _get_osfhandle().

This fixes node.js unit tests test-fs-read-stream.js and
test-listen-fd-ebadf.js.

### CODE REVIEW

File: src/win/handle-inl.h [hunk 1]
```diff
@@ -161,4 +162,18 @@ INLINE static void uv_process_endgames(uv_loop_t* loop) {
   }
 }
 
+INLINE static HANDLE uv_get_osfhandle(int fd)
```
Comment (tjfontaine) [inline src/win/handle-inl.h]: If this is purely internal I think the convention is `uv__`

**Comment by orangemocha** [comment]
@tjfontaine Updated to uv__get_osfhandle()

File: src/win/core.c [hunk 2]
```diff
@@ -41,6 +42,33 @@
 static uv_once_t uv_default_loop_init_guard_ = UV_ONCE_INIT;
 
 
+#ifdef _DEBUG
+/* Our crt debug report handler allows us to temporarily disable asserts */
+/* just for the current thread. */
+
+__declspec( thread ) BOOL uv__crt_assert_enabled = TRUE;
```
Comment (piscisaureus) [inline src/win/core.c]: prefer to use standard c types and not windows types if you can help it. I think in this case using an int and setting it to 1 would suffice?

File: src/win/core.c [hunk 3]
```diff
@@ -41,6 +42,33 @@
...
+
+static int uv__crt_dbg_report_handler(int report_type, char *message, int *ret_val) {
+  if (uv__crt_assert_enabled || report_type != _CRT_ASSERT)
+    return FALSE;
+  
+  if (ret_val)
+  {
```
Comment (piscisaureus) [inline src/win/core.c]: nit: style, put the { on the same line as the if statement.

File: src/win/internal.h [hunk 4]
```diff
@@ -35,6 +35,25 @@
 # define INLINE inline
 #endif
 
+
+#ifdef _DEBUG
+extern __declspec( thread ) BOOL uv__crt_assert_enabled;
+
+#define UV_BEGIN_DISABLE_CRT_ASSERT()                           \
+  {                                                             \
+    BOOL uv__saved_crt_assert_enabled = uv__crt_assert_enabled;  \
```
Comment (piscisaureus) [inline src/win/internal.h]: backslash alignment looks off

File: src/win/internal.h [hunk 5]
```diff
@@ -35,6 +35,25 @@
...
+    uv__crt_assert_enabled = FALSE;
+  
+
+#define UV_END_DISABLE_CRT_ASSERT()                             \
+    uv__crt_assert_enabled = uv__saved_crt_assert_enabled;       \
```
Comment (piscisaureus) [inline src/win/internal.h]: ditto

**Comment by piscisaureus** [comment]
Excellent, we were in dire need for a good solution for this problem. Thanks!

A couple of small nits. 
Also I think this may break the mingw build, but that can be fixed later when/if people complain about it.

File: src/win/core.c (hunk 2 above)
Comment (orangemocha) [inline src/win/core.c]: no problem. I hope the TRUE/FALSE defines are still acceptable

**Comment by orangemocha** [comment]
I believe all your suggested corrections have been incorporated. Thank you!

**Comment by piscisaureus** [comment]
Thanks, landed in c0716b3.

**Pull request closed without merging**